File: 프로그래머스/오픈채팅방.py
# https://programmers.co.kr/learn/courses/30/lessons/42888

def solution(record):
    user = {}
    result = []
    for rec in record:
        command = rec.split(' ')
        if command[0] == 'Enter':
            user[command[1]] = command[2]
        elif command[0] == 'Change':
            user[command[1]] =command[2]

    for rec in record:
        command = rec.split(' ')
        if command[0] == 'Enter':
            result.append(user[command[1]]+'님이 들어왔습니다.')
        elif command[0] == 'Leave':
            result.append(user[command[1]]+'님이 나갔습니다.')
        else :
            continue
    return result

def solution2(record):
    user = {}
    result = []
    print = {'Enter':'님이 들어왔습니다.', 'Leave':'님이 나갔습니다.'}
    for rec in record:
        command = rec.split()
        if command[0] in ['Enter','Change']:
            user[command[1]] =command[2]

    for rec in record:
        command = rec.split()
        if command[0] == 'Enter':
            result.append(user[command[1]]+'님이 들어왔습니다.')
        elif command[0] == 'Leave':
            result.append(user[command[1]]+'님이 나갔습니다.')
    return result

File: Programmers-coding-test/Level 2/오픈채팅방.js
// https://programmers.co.kr/learn/courses/30/lessons/42888

function solution(record) {
  var answer = [];
  const user = {};
  for (rec of record) {
    const [command, id, nickname] = rec.split(' ');
    if (command == 'Enter' || command == 'Change') {
      user[id] = nickname;
    }
  }
  for (rec of record) {
    const [command, id, nickname] = rec.split(' ');
    if (command == 'Enter') {
      answer.push(`${user[id]}님이 들어왔습니다.`);
    } else if (command == 'Leave') {
      answer.push(`${user[id]}님이 나갔습니다.`);
    }
  }
  return answer;
}

function solution2(record) {
  const user = {};
  const action = [];
  const print = { Enter: '님이 들어왔습니다.', Leave: '님이 나갔습니다.' };
  record.forEach((rec) => {
    const [command, id, nickname] = rec.split(' ');
    // Enter, Leave
    if (command != 'Change') {
      action.push([command, id]);
    }
    // Change
    if (nickname) {
      user[id] = nickname;
    }
  });
  return action.map(([command, id]) => {
    return `${user[id]}${print[command]}`;
  });
}
